lib/test/t_HSICEstimatorConditionalSensitivity_std.cxx

```diff
@@ -0,0 +1,144 @@
+  RandomGenerator::SetSeed(0);
+  try
+  {
+    /* Definition of the marginals */
+    Uniform X1(-M_PI, M_PI);
+    Uniform X2(-M_PI, M_PI);
+    Uniform X3(-M_PI, M_PI);
+
+    ComposedDistribution::DistributionCollection aCollection;
+    aCollection.add(X1);
+    aCollection.add(X2);
+    aCollection.add(X3);
+
+    /* 3d distribution made with independent marginals */
+    Distribution distX  = ComposedDistribution(aCollection);
+
+    /* Get a sample of it */
+    SignedInteger size = 100;
+    Sample X = distX.getSample(size);
+
+    /* The Ishigami model */
+    Description input = {"X1", "X2", "X3"};
+    SymbolicFunction modelIshigami(input, Description(1, "sin(X1) + 5.0 * (sin(X2))^2 + 0.1 * X3^4 * sin(X1)"));
+
+    /* Apply model: Y = m(X) */
+    Sample Y = modelIshigami(X);
+
+    /* We define the covariance models for the HSIC indices. */
+    /* For the input, we consider a SquaredExponential covariance model. */
+    HSICEstimator::CovarianceModelCollection covarianceList(0);
+
+    /* Input sample */
+    for (UnsignedInteger i = 0; i < 3; ++i)
+    {
+      Sample Xi = X.getMarginal(i);
+      CovarianceModel Cov;
+      Cov = SquaredExponential(1);
+      Cov.setScale(Xi.computeStandardDeviation());
+      covarianceList.add(Cov);
+    }
+
+    /* Output sample with squared exponential covariance */
+    CovarianceModel Cov2;
+    Cov2 = SquaredExponential(1);
+    Cov2.setScale(Y.computeStandardDeviation());
+    covarianceList.add(Cov2);
+
+    /* We choose an estimator type :
+       - unbiased: HSICUStat (not available here!!);
+       - biased: HSICVStat.
+    */
+    HSICVStat estimatorType;
+
+    /* We define a distance function for the weights
+       For the CSA, the critical domain is [5,+inf].
+
+       We set the interval as [5,1000] and make the upper bound
+       not finite.
+    */
+    Interval interval(5, 1000);
+    Interval::BoolCollection bcoll;
+    bcoll.add(false);
+    interval.setFiniteUpperBound(bcoll);
+    Function g = DistanceToDomainFunction(interval);
+
+    /* We create the weight function as a composition */
+    Scalar stdDev = Y.computeStandardDeviation()[0];
+    Description inVar(2);
+    inVar[0] = "x";
+    inVar[1] = "s";
+    SymbolicFunction foo(inVar, Description(1, "exp(-x/s)"));
+    Indices set(1);
+    set[0] = 1;
+    Bool parametersSet = true;
+    Point referencePoint(1, 0.1 * stdDev);
+    MemoizeFunction fooh(foo);
+    ParametricFunction g2(fooh, set, referencePoint, parametersSet);
+
+    /* The weight function */
+    Function weight = ComposedFunction(g2, g );
+
+    /* We eventually build the HSIC object! */
+    HSICEstimatorConditionalSensitivity CSA(covarianceList, X, Y, estimatorType, weight);
+
+    /* We get the R2-HSIC */
+    Point referenceR2HSIC = {0.0371735, 0.0052413, 0.235519};
+    Point R2HSIC = CSA.getR2HSICIndices();
+    //œstd::cout << "R2-HSIC: " << R2HSIC << std::endl;
+    assert_almost_equal(R2HSIC, referenceR2HSIC);
```

Comment:
It would be better if you specified the relative and absolute tolerances here and elsewhere. In this case, I would go with relative tolerance 1e-3 and absolute tolerance 0:
```suggestion
    assert_almost_equal(R2HSIC, referenceR2HSIC, 1e-3, 0.0);
```
The ability to specify tolerances is why I asked you to remove prints.

Reply:
On removing the print (and the string comparison of results) I agree. But what's wrong with default values?

Reply:
I do not like the default tolerances (and, in practice, I can't remember them ever being used before in OT):

https://github.com/openturns/openturns/blob/07ba0debc4b41b44c7858b685039562ab36a7187/lib/include/openturns/OTtestcode.hxx#L299

There is no reason to have non-zero absolute tolerance if the reference value is non-zero. If, say, the reference value is 1e-10 and the computed value 1e-9, we should consider it wrong but the default absolute tolerance (1e-8) would make it acceptable.

Reply:
I still disagree. An absolute error measures the difference and nothing more. A zero-tolerance is OK for exact values but for floating points the 1e-8 default is OK.

Reply:
This is not very important in the context of HSIC estimators (so feel free to ignore it) because values close to zero do not need to be too precise. It is more a matter of general principle.

> An absolute error measures the difference and nothing more.

That is precisely my problem. Differences do not mean much by themselves, only *relative* errors do.

> A zero-tolerance is OK for exact values

The testing formula is https://github.com/openturns/openturns/blob/07ba0debc4b41b44c7858b685039562ab36a7187/lib/include/openturns/OTtestcode.hxx#L303

Having zero absolute tolerance means that this is reduced to

``
 if (std::abs(a - b) > rtol * std::abs(b))
``

which is a sensible formula whenever the reference value `b` is non-zero. "Zero absolute tolerance" is only equivalent to "zero tolerance" when `b==0`. 

>  but for floating points the 1e-8 default is OK.

1e-8 is 300 orders of magnitude larger than the smallest normal floating point value (about 2.2e-308). HSIC estimators may be unlikely to take such extremely low values, but other quantities manipulated in OT do (for example logPDFs manipulated in MCMC classes).

Tagging @mbaudin47 who may have further insights on this topic.

Reply:
@josephmure 1e-8 is a default value for testing differences between values of order 1, for which absolute and relative errors are mostly the same. You will probably *never* manipulate values as small as 2.2e-308 as in these ranges, you work on a logarithmic scale and get values much more of order -700